File: commands/Utility/image.js
const { Command } = require("klasa");
const { MessageEmbed } = require("discord.js");
const { random, slice } = require("../../utils/utils.js");
const ladybug = require("ladybug-fetch");
const cheerio = require("cheerio");

class Image extends Command {
  constructor(...args) {
    super(...args, {
      description: "Searches for images from Bing.",
      usage: "<query:string>",
      aliases: ["imagesearch", "img", "imgsearch"],
      cooldown: 3,
      extendedHelp: "Use --index=<n> to retrieve nth image, by default it picks a random one everytime."
    });
    this.userAgent = { "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2228.0 Safari/537.36" };
    this.url = (query, nsfw) => `https://www.bing.com/images/search?q=${query}&view=detailv2&safeSearch=${nsfw ? "off" : "strict"}`;
  }

  async run(msg, [query]) {
    // Quick validation before we go straight to requesting.
    if(msg.flags.index && isNaN(parseInt(msg.flags.index)))
      throw "--index Provided but not a number.";
    const { text, status } = await ladybug(this.url(query, msg.channel.nsfw))
      .set(this.userAgent);
    if(status !== 200) throw "Something went wrong with Bing.";
    const $ = cheerio.load(text);
    const results = [];
    $(".item a[class=\"thumb\"]").each((_, el) => {
      const details = $(el).parent().find(".fileInfo").text().split(" ");
      return results.push({
        url: $(el).attr("href"),
        width: details[0],
        height: details[2],
        format: details[3],
        size: details[4]
      });
    });
    const index = msg.flags.index ? parseInt(msg.flags.index) : undefined;
    if(index && index > results.length)
      throw `--index provided as \`${index}\` but there is only \`${results.length}\` results.`;
    const image = index ? results[index - 1] : random(results);
    const embed = new MessageEmbed()
      .setTitle(`Image Results: ${slice(query, 128)}`)
      .setImage(image.url)
      .setFooter(`Size: ${image.size}, Resolution: ${image.width}x${image.height}, Format: ${image.format}`)
      .setColor(0xFF0000);
    return msg.send({ embed });
  }
}

module.exports = Image;

File: commands/utility/image.go
package utility

import (
  "net/http"
  "net/url"
  "github.com/PuerkitoBio/goquery"
  "fmt"
  "strings"
  "github.com/pollen5/taiga/utils"
  "github.com/sapphire-cord/sapphire"
  "strconv"
)

// Get an image from Bing.
// Usage: <query:string...>
// Aliases: imgsearch, imagesearch, searchimage, searchimg, img
// Cooldown: 5
func Image(ctx *sapphire.CommandContext) {
  type Image struct {
    URL string
    Width string
    Height string
    Format string
    Size string
  }
  var safe string = "strict"
  if ctx.Channel.NSFW { safe = "off" }
  uri := url.Values{}
  uri.Add("q", ctx.JoinedArgs())
  uri.Add("view", "detailv2")
  uri.Add("safeSearch", safe)
  req, _ := http.NewRequest(http.MethodGet, "https://www.bing.com/images/search?" + uri.Encode(), nil)
  req.Header.Set("User-Agent", "Lynx/2.8.8rel.2 libwww-FM/2.14 SSL-MM/1.4.1 OpenSSL/1.0.2l")
  res, err := http.DefaultClient.Do(req)
  if err != nil {
    ctx.Reply("Error: %s\n", err)
    return
  }
  defer res.Body.Close()
  if res.StatusCode != 200 {
    ctx.Reply("Bing responded with non-200: %d %s", res.StatusCode, res.Status)
    return
  }
  doc, err := goquery.NewDocumentFromReader(res.Body)
  if err != nil {
    ctx.Reply("Error parsing HTML: %s", err)
    return
  }
  results := []*Image{}
  doc.Find(".item a[class=\"thumb\"]").Each(func(_ int, s *goquery.Selection) {
    details := strings.Split(s.Parent().Find(".fileInfo").Text(), " ")
    url, _ := s.Attr("href")
    results = append(results, &Image{
      URL: url,
      Width: details[0],
      Height: details[2],
      Format: details[3],
      Size: details[4],
    })
  })
  if len(results) < 1 {
    ctx.Reply("No Results Found.")
    return
  }
  flag, ok := ctx.Flags["index"]
  var index int = 0
  if ok {
    num, err := strconv.Atoi(flag)
    if err != nil {
      ctx.Reply("--index Provided but not a number.")
      return
    }
    index = num - 1
    if num > len(results) || num < 0 {
      ctx.Reply("Invalid range for --index, there is only %d entries.", len(results))
      return
    }
  } else {
    index = utils.RandNumber(len(results))
  }
  img := results[index]
  ctx.BuildEmbed(sapphire.NewEmbed().SetColor(0xDFAC7C).
    SetTitle("Image Results").
    SetImage(img.URL).
    SetFooter(fmt.Sprintf("Size: %s, Resolution: %sx%s, Format: %s",
      img.Size, img.Width, img.Height, img.Format)))
}
